source/server/drain_manager_impl.h

```diff
@@ -21,23 +25,41 @@ namespace Server {
  */
 class DrainManagerImpl : Logger::Loggable<Logger::Id::main>, public DrainManager {
 public:
-  DrainManagerImpl(Instance& server, envoy::config::listener::v3::Listener::DrainType drain_type);
+  DrainManagerImpl(Instance& server, envoy::config::listener::v3::Listener::DrainType drain_type,
+                   Event::Dispatcher& dispatcher);
 
   // Network::DrainDecision
   bool drainClose() const override;
+  Common::CallbackHandlePtr addOnDrainCloseCb(DrainCloseCb cb) const override;
 
   // Server::DrainManager
   void startDrainSequence(std::function<void()> drain_complete_cb) override;
   bool draining() const override { return draining_; }
   void startParentShutdownSequence() override;
+  DrainManagerPtr
+  createChildManager(Event::Dispatcher& dispatcher,
+                     envoy::config::listener::v3::Listener::DrainType drain_type) override;
+  DrainManagerPtr createChildManager(Event::Dispatcher& dispatcher) override;
 
 private:
+  void addDrainCompleteCallback(std::function<void()> cb);
+
   Instance& server_;
+  Event::Dispatcher& dispatcher_;
   const envoy::config::listener::v3::Listener::DrainType drain_type_;
 
   std::atomic<bool> draining_{false};
   Event::TimerPtr drain_tick_timer_;
   MonotonicTime drain_deadline_;
+  mutable Common::CallbackManager<std::chrono::milliseconds> cbs_{};
+  std::vector<std::function<void()>> drain_complete_cbs_{};
```

Comment:
For my understanding, how can there ever be more than a single drain complete callback? It seems that now startDrainSequence can be called multiple times? In your next follow up can you add more comments on how/why this happens?

Reply:
I will add a comment.

The gist is that, because drain managers may start draining independently of their parent, it is possible that components can "race" in draining. Whereas before local draining was handled by independent state changes that were reflected through the DrainDecision interface.

Reply:
Added in e568d7b907712df3e034cc65887bbc185ea51563 on #16201